accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+level and the level of compiled modules, but would fall a little short when it
+comes to facilitating debugging interpreted languages. It would also require
+that users interact with the debugger using gdb’s native interface or through a
+Debug Adapter Protocol interface built on top of gdb requiring additional setup
+and configuration. We would need to build additional abstractions on top of gdb
+to support the same sort of interface that we’ve described for supporting
+interacting with SpiderMonkey’s debugger, and at that point we would be
+complicating our ability to support time-travel debugging or record+replay in
+the future by switching from debugging Wasm to debugging native code.
+
+Choosing an existing debugger like gdb or lldb would compromise our three
+principles outlined in the [motivation](#motivation) section &mdash;
+particularly principle (2) and supporting interpreted debuggees[^possible]
+&mdash; and therefore we feel that they are not viable choices.
+
+[^possible]: We think it is *probably*, *technically* possible to implement
+debugging interpreted programs in `gdb` or `lldb` through their plugin
+interfaces. However, we believe that the user experience would be somewhat
+clunky and the implementation difficult.
+
+## Breakpoints
+
+Instead of inserting calls to a utility function to check for breakpoints set at
+a specific instruction, we could insert `nop` instructions. Those `nop`
+instructions would be overwritten with `int3` instructions to generate `SIGTRAP`
+when executed, allowing us to then make a decision about either continuing if
+the execution was from an unrelated instance, or actually pausing execution and
+yielding control to the debugger.
+
+One significant complication here would be passing the metadata required by the
+breakpoint handler is no longer as simple: we would need to recover that
+information in the signal handler instead of having it passed into the utility
+function called between each instruction. Additionally, this would complicate
+the signal handler for Wasmtime.
+
+## Watchpoints
+
+An alternative to the approach for implementing memory watchpoints above is to
+use a similar mechanism to what we use for eliding bounds checks: we can have
+virtual memory mapped in with write-permissions off. Then, any time the guest
+program tries to write to that page, a signal will be raised. Our signal handler
+will then check if the address being written to is in our watch-list. If yes,
+then pause execution and wait for user input. If not, then we perform the write
+and resume. This approach introduces similar new complexity into our signal
+handler as the alternative implementation for breakpoints, however it also
+avoids the additional overhead of calling into a local utility function for each
+memory write. Similar to the breakpoint implementation outlined above, it would
+also require recovering the context for a watchpoint through a mechanism other
+than function arguments.
```

Comment:
Aha, I made this suggestion above. I should've scrolled down before writing it! :)